App insights

Smarker opened the pull request:
Added appinsights to my code from #6. Please let me know if there are any other areas where you think appinsights should be used or if you have suggestions for a different approach.

Added **APPINSIGHTS_INSTRUMENTATIONKEY** to env var.

jcjimenez commented on src/storageClients/BlobStorageFactsManager.js:
```diff
@@ -9,114 +9,114 @@ const FACTS_STORAGE_CONNECTION_STRING = process.env.FACTS_STORAGE_CONNECTION_STR
+  GetFact: function (id, callback, logger) { // eslint-disable-line no-unused-vars
+    var blobSvc = azure.createBlobService(FACTS_STORAGE_CONNECTION_STRING);
+    let blobPrefix = id.split('-')[0] + '/' + id.split('-')[1] + '/' + id.split('-')[2];
+    let factIndex = id.split('-')[3];
```
I know you didn't change this, but do you know if we want to do something other than throw on a potential invalid index from a malformed id?

c-w commented:
Given that this is based on #6, I've temporarily changed the base of this review branch to #6's branch so that the changes in this pull request are made more obvious. There are a few ways forward from here:

1) After getting sufficient sign-offs, change the base to master again.
2) After getting a sign-off, merge the changes into #6.
3) Wait for #6 to be merged and then merge this.

c-w commented on src/connectors/CassandraConnector.js:
```diff
@@ -16,22 +25,35 @@ module.exports = {
-        .catch(reject);
+        .then(() => {
+          let duration = Date.now() - start;
+          iclient.trackMetric('Cassandra shutdown time', duration);
```
I wonder if we can make this more generic/re-usable. Pseudo-code:

```js
withMetric('Cassandra connection time', () => new Promise((resolve, reject) => {
  client.connect()
  .catch((err) => {
    client.shutdown();
	reject(err);
  });
});

function withMetric(metricName, promiseAction) {
  let start = Date.now();
  promiseAction().then((val) => {
    let duration = Date.now() - start;
	iclient.trackMetric(metricName, duration);
	resolve(val);
  })
  .catch((err) => {
    iclient.trackException(err)
	reject(err);
  });
}
```

Adding a slight level of abstraction on top of using the naked client also has the advantage of separating the logging implementation from the code that does the business logic. Here, everything is still fine, but it's a slippery slope and easy to get to the situation where the logging/argument-checking code takes up the majority of a method and drowns out the business logic parts. Using this decorator style to handle logging will set a good example to prevent that.

Maybe it's worth spending 15 minutes to see if we can factor out the logging code like this, but it's not a high priority item so don't worry if you run into a roadblock.

Smarker replied:
good idea, I will try to do this

c-w commented on src/resolvers/Topics.js:
```diff
@@ -1,21 +1,64 @@
 'use strict';
+const appInsights = require('applicationinsights');
+appInsights.setup(process.env.APPINSIGHTS_INSTRUMENTATIONKEY)
```
We have this code repeated in quite a few places. Would it be possible to encapsulate this in one place to reduce the duplication (e.g. `src/loggingClient`)?

This has the added benefit that we can introduce a null-object ApplicationInsights client if the instrumentation key is not set so that we can enable running the app locally without requiring the instrumentation key to be specified which is a better developer experience.

Pseudo-code:

```js
const ikey = process.env.APPINSIGHTS_INSTRUMENTATIONKEY

let client;
if (ikey) {
  appInsights.setup(ikey);
  client = appInsights.getClient();
} else {
  client = {
    trackException: (ex) => console.log("exception " + ex),
    trackEvent: (route, details) => console.log("event " + route + " " + JSON.stringify(details))
  };
}

module.exports = {
  client: client
};
```

Smarker replied:
yes, good idea

c-w commented on src/connectors/CassandraConnector.js:
```diff
@@ -1,10 +1,19 @@
 'use strict';
+const appInsights = require('applicationinsights');
+appInsights.setup(process.env.APPINSIGHTS_INSTRUMENTATIONKEY)
```
Please also document the key in the README.

erikschlegel replied:
This is unnecessary as the  `applicationinsights` package automatically loads the instance based on the `APPINSIGHTS_INSTRUMENTATIONKEY` env variable https://github.com/Microsoft/ApplicationInsights-node.js#getting-started

Smarker replied:
ok!

erikschlegel commented on src/connectors/CassandraConnector.js:
```diff
@@ -1,10 +1,19 @@
 const CASSANDRA_CONTACT_POINTS = process.env.CASSANDRA_CONTACT_POINTS;
-
 const options = {
   contactPoints: [CASSANDRA_CONTACT_POINTS],
```
You should check that the `CASSANDRA_CONTACT_POINTS ` env variable is defined before it's used, and raise an app insight exception.

erikschlegel commented on src/connectors/CassandraConnector.js:
```diff
@@ -1,10 +1,19 @@
 'use strict';
+const appInsights = require('applicationinsights');
+appInsights.setup(process.env.APPINSIGHTS_INSTRUMENTATIONKEY)
+  .setAutoDependencyCorrelation(true)
```
I believe these all are enabled by default

Smarker commented:
I edited these files to address the comments for this PR:
* connectors/CassandraConnector -> executeQueries
* resolvers/Settings.js -> createOrReplaceSite (using insertSeedTopics)
* storageClients/BlobStorageManager -> getTopicsBySiteType
* storageClients/CassandraTableStorageManager -> prepareInsertTopic, batch (using chunkQueryPromises)

I addressed the comments by
* removing all instances of bluebird
* adding a connection pooling option to the Cassandra client
* making Cassandra client a singleton
* creating getTopicsBySiteType using a http request rather than BLOB_STORAGE_CONNECTION_STRING
* editing the insert topics statement
* adding app insights within the functions to track

erikschlegel commented on src/connectors/CassandraConnector.js:
```diff
@@ -1,37 +1,54 @@
+  keyspace: process.env.CASSANDRA_KEYSPACE,
+  pooling: {
+    coreConnectionsPerHost: {
+      [distance.local] : 10,
```
How did you come up with these settings?

erikschlegel commented on src/connectors/CassandraConnector.js:
```diff
@@ -1,37 +1,54 @@
 'use strict';
+const appInsights = require('applicationinsights');
+appInsights.setup(process.env.APPINSIGHTS_INSTRUMENTATIONKEY).start();
```
re: my earlier comment, setting `APPINSIGHTS_INSTRUMENTATIONKEY` is redundant as the `applicationinsights` library automatically does this for you

erikschlegel commented on src/connectors/CassandraConnector.js:
```diff
@@ -1,37 +1,54 @@
-
-const CASSANDRA_CONTACT_POINTS = process.env.CASSANDRA_CONTACT_POINTS;
-
+/** Cassandra Client Options
```
Can you comment your code a bit more so other developers have more context into how this driver works(i.e. connection pooling, etc)

erikschlegel commented on src/connectors/CassandraConnector.js:
```diff
@@ -1,37 +1,54 @@
+   * @param String eventIdentifier  A string to identify the event for appinsights
+   * @return {Promise}
+   */
+  executeQueries: (queries, eventIdentifier) => {
```
I'm confused, why would one want to execute a batch of queries? Not sure if this is meant to address a batch of mutation statements(`insert`, `update`). If yes, please name your functions and variables in a way that helps other developers better understand your code.

erikschlegel commented on src/connectors/CassandraConnector.js:
```diff
@@ -1,37 +1,54 @@
+      let numQueries = queries.length;
+      cassandraTableStorageManager.batch(client, queries)
+        .then(res => {
+          iclient.trackEvent(eventIdentifier, {numQueries: numQueries});
```
I think this will be tracked in app insights a key value bag of `properties` and not `measurements`. Is that your intention here?

erikschlegel commented on src/connectors/CassandraConnector.js:
```diff
@@ -1,37 +1,54 @@
+      cassandraTableStorageManager.batch(client, queries)
+        .then(res => {
+          iclient.trackEvent(eventIdentifier, {numQueries: numQueries});
+          resolve(res);
```
How does the connection get returned to the pool and/or closed?

erikschlegel commented on src/connectors/CassandraConnector.js:
```diff
@@ -1,37 +1,54 @@
+/** Code should share the same Client instance across the application
+ * http://docs.datastax.com/en/developer/nodejs-driver/3.2/coding-rules/
+ */
+const client = new cassandra.Client(options);
```
Can this throw an exception if none of the Contact points are available? Not sure if it does or not.

erikschlegel commented on src/resolvers/Settings.js:
```diff
@@ -31,23 +30,27 @@ module.exports = {
 
     return new Promise((resolve, reject) => {
```
Where is `Promise` imported?

erikschlegel commented on src/resolvers/Settings.js:
```diff
@@ -31,23 +30,27 @@ module.exports = {
 
```
Lets create a new resolver called `CreateNewSite` and rename `createOrReplaceSite`->`ReplaceSite` and refactor accordingly. Both resolvers have two very different code paths and the code quality is starting to suffer because of it.

erikschlegel replied:
You'll also need to do some checks in the event the `CreateNewSite` is called on an already existing site, which in this case would continually add duplicate topics.

erikschlegel commented on src/storageClients/BlobStorageManager.js:
```diff
@@ -1,75 +1,36 @@
-          resolve(blobNames);
-        })
-        .catch(reject);
+      const url = `https://fortiscentral.blob.core.windows.net/settings/siteTypes/${siteType}/topics/defaultTopics.json`;
```
the root context of the central host should be coming from the environment and not hard coded, as this is already available via the deployment.

erikschlegel replied:
you should be performing some validation on `siteType`

erikschlegel commented on src/resolvers/Settings.js:
```diff
@@ -292,26 +295,24 @@ module.exports = {
+    blobStorageManager.getTopicsBySiteType(siteType)
+      .then(topics => {
+        let queries = [];
+        for(let topic of topics) {
```
Im not sure I fully understand how this is working, as the json schema in the blob containers dont match up with how you're binding `topics` to the cassandra prepared statement ```{"id": 62, "topic": "climate", "value": "#climate"}, {"id": 63, "topic": "climate", "value": "climatechange"}, {"id": 64, "topic": "climate", "value": "globalwarming"}```

erikschlegel replied:
Also, I dont see any unit tests in your PR, which I would assume would catch this as from the surface this looks like it would fail.

erikschlegel commented on src/storageClients/BlobStorageManager.js:
```diff
@@ -1,75 +1,36 @@
-const BLOB_STORAGE_CONNECTION_STRING = process.env.BLOB_STORAGE_CONNECTION_STRING;
-const TOPIC_SEED_CONTAINER = process.env.TOPIC_SEED_CONTAINER;
+const appInsights = require('applicationinsights');
+appInsights.setup(process.env.APPINSIGHTS_INSTRUMENTATIONKEY).start();
```
Please remove `APPINSIGHTS_INSTRUMENTATIONKEY` reference. https://github.com/Microsoft/ApplicationInsights-node.js#getting-started

erikschlegel commented on src/storageClients/BlobStorageManager.js:
```diff
@@ -1,75 +1,36 @@
+
+      request(GET, (error, response, body) => {
+        if (!error && response.statusCode === 200 && body) {
+          iclient.trackEvent(`GET ${url}`);
```
the event name should be something thats short and consistent(`REQ_SITE`). You should be adding the url as a key entry in the `properties` map.

erikschlegel commented on src/storageClients/BlobStorageManager.js:
```diff
@@ -1,75 +1,36 @@
+          iclient.trackEvent(`GET ${url}`);
+          resolve(body);
+        } else {
+          const errMsg = `[${error}] occured while fetching request for url ${url}`;
```
I'd also log the ` response.statusCode` in your `errMsg`

erikschlegel commented on src/storageClients/CassandraTableStorageManager.js:
```diff
@@ -1,15 +1,34 @@
 module.exports = {
 
-  prepareInsertTopic: topic => { //TODO: the insert statement to reflect the final topic schema
+  prepareInsertTopic: topic => {
```
This insert statement doesnt belong in the table storage manager. Perhaps have a separate file to maintain your queries / mutation  statements.

erikschlegel commented on src/storageClients/CassandraTableStorageManager.js:
```diff
@@ -1,15 +1,34 @@
 const BATCH_LIMIT = 10;
 
 module.exports = {
 
```
It's a bit difficult to follow the `chunkQueryPromises ` method. I mentioned using the `async` forEachLimit` in an earlier comment. Would probably help simplify things a bit.

erikschlegel commented on src/storageClients/CassandraTableStorageManager.js:
```diff
@@ -1,15 +1,34 @@
+      ]
     };
   },
 
```
the function name of `batch` seems a bit vague.

erikschlegel commented on src/storageClients/CassandraTableStorageManager.js:
```diff
@@ -1,15 +1,34 @@
     return {
-      query: 'INSERT INTO topics (id, topic, value) VALUES (?, ?, ?)',
-      params: [ topic.id, topic.topic, topic.value ]
+      query: `INSERT INTO computedtopics (
```
spark should be the only process writing to this table. I think you meant to write to `watchlist`

erikschlegel commented on test/CassandraConnector.js:
```diff
@@ -1,25 +1,28 @@
+    it('should resolve the promise', function() {
+      let queries = [
+        {
+          query: 'INSERT INTO Topics (id, topic, value) VALUES (?, ?, ?)',
```
there is no table called `Topics` https://github.com/CatalystCode/fortisdeploy/blob/119d3d847826b2cbd45791b3cf41c2ce1d60144d/ops/storage-ddls/cassandra-setup.cql
